File: aria-backend/src/aria/connectors/neo4j.ts
import neo4j, { Driver } from "neo4j-driver";

import { ariaConfig, isConnectorLive } from "../config";
import { mockDependencyGraph } from "../mock-data";
import { DependencyGraphResult } from "../types";

let cachedDriver: Driver | null = null;

function getDriver(): Driver {
  if (cachedDriver) {
    return cachedDriver;
  }

  cachedDriver = neo4j.driver(
    ariaConfig.neo4j.uri ?? "",
    neo4j.auth.basic(ariaConfig.neo4j.username ?? "", ariaConfig.neo4j.password ?? ""),
  );

  return cachedDriver;
}

function toStringArray(values: unknown[]): string[] {
  return Array.from(
    new Set(
      values
        .map((value) => (typeof value === "string" ? value : null))
        .filter((value): value is string => Boolean(value)),
    ),
  );
}

export class Neo4jConnector {
  async fetchBlastRadius(service: string): Promise<DependencyGraphResult> {
    if (!isConnectorLive("neo4j")) {
      return mockDependencyGraph(service);
    }

    const driver = getDriver();
    const session = driver.session({ database: ariaConfig.neo4j.database });

    try {
      const downstreamResult = await session.run(
        `
        MATCH (down:Service)-[:DEPENDS_ON*1..2]->(s:Service {name: $service})
        RETURN DISTINCT down.name AS serviceName
        LIMIT 25
        `,
        { service },
      );

      const upstreamResult = await session.run(
        `
        MATCH (s:Service {name: $service})-[:DEPENDS_ON*1..2]->(up:Service)
        RETURN DISTINCT up.name AS serviceName
        LIMIT 25
        `,
        { service },
      );

      const impacted = toStringArray(
        downstreamResult.records.map((record) => record.get("serviceName")),
      );
      const upstream = toStringArray(
        upstreamResult.records.map((record) => record.get("serviceName")),
      );

      return {
        impactedServices: impacted,
        upstreamServices: upstream,
        connectorMode: "live",
      };
    } catch {
      return mockDependencyGraph(service);
    } finally {
      await session.close();
    }
  }
}

File: aria-backend/aria/connectors/neo4j.py
"""Neo4j connector — sync driver for use inside Strands tools."""
from __future__ import annotations

import logging

from aria.config import config
from aria.mock_data import mock_dependency_graph
from aria.types import DependencyGraphResult

logger = logging.getLogger(__name__)

_driver = None


def _get_driver():
    global _driver
    if _driver is None:
        from neo4j import GraphDatabase

        _driver = GraphDatabase.driver(
            config.neo4j_uri or "",
            auth=(config.neo4j_username or "neo4j", config.neo4j_password or ""),
        )
    return _driver


def fetch_blast_radius(service: str) -> DependencyGraphResult:
    if not config.connector_live("neo4j"):
        return mock_dependency_graph(service)

    driver = _get_driver()
    try:
        with driver.session(database=config.neo4j_database) as session:
            downstream = session.run(
                """
                MATCH (down:Service)-[:DEPENDS_ON*1..2]->(s:Service {name: $service})
                RETURN DISTINCT down.name AS serviceName
                LIMIT 25
                """,
                service=service,
            )
            upstream = session.run(
                """
                MATCH (s:Service {name: $service})-[:DEPENDS_ON*1..2]->(up:Service)
                RETURN DISTINCT up.name AS serviceName
                LIMIT 25
                """,
                service=service,
            )
            impacted = list({r["serviceName"] for r in downstream if r["serviceName"]})
            up_list = list({r["serviceName"] for r in upstream if r["serviceName"]})

        return DependencyGraphResult(
            impactedServices=impacted,
            upstreamServices=up_list,
            connectorMode="live",
        )
    except Exception as exc:
        logger.warning("Neo4j query failed (%s), using mock.", exc)
        return mock_dependency_graph(service)
